autoload/phpactor/project/bufferMatcher.vim

```diff
@@ -0,0 +1,129 @@
+function! phpactor#project#bufferMatcher#assignFileToProject(filename)
+  if !exists('g:phpactorLoaded')
+    return
+  endif
+
+  if exists('b:project')
+    return
+  endif
+  let l:project = g:phpactorBufferMatcher.matchFileToProject(a:filename)
+  if v:null is l:project
+    return
+  endif
+  let b:project = l:project
+  if g:phpactorBufferMatcher.getRepository().hasProject(b:project)
+    return
+  endif
+
+  call g:phpactorBufferMatcher.getRepository().addProject(b:project)
+endfunction
+
+function! phpactor#project#bufferMatcher#create(projectRepository, projectRootMarkers, filesystemRootMarkers, initialCwd) abort
+  let l:filesystemRootMarkers = a:filesystemRootMarkers
+  if index(l:filesystemRootMarkers, '/') < 0
+    call add(l:filesystemRootMarkers, '/')
+  endif
+
+  return {
+        \ 'initialCwd': a:initialCwd,
+        \ 'projectRootMarkers': a:projectRootMarkers,
+        \ 'filesystemRootMarkers': l:filesystemRootMarkers,
+        \ 'repository': a:projectRepository,
+        \ 'getRepository': function('s:getRepository'),
+        \ 'matchFileToProject': function('s:matchFileToProject')
+        \ }
+endfunction
+
+function s:matchFileToProject(file) dict abort
+  let l:project = self.repository.findProjectContainingFile(a:file)
+
+  if type(l:project) == type({})
+    return l:project
+  endif
+
+  " early version to refactor or change
+  " heredoc are available since vim 8.1.1354
+  let l:message = printf(join([
+        \ 'Phpactor [RPC]',
+        \ 'There is no project enabled for "%s" yet.',
+        \ 'Select the way of assign the project root',
+        \ 'If cancelled, "%s" will be used.'
+        \], "\n"), a:file, self.initialCwd)
+
+  " see phpactor#_path
+  let l:initialDirectory = fnamemodify(a:file, ':p:h')
+  let l:rootDirByMarker = s:searchDirectoryUpwardForRootMarkers(
+        \ l:initialDirectory,
+        \ self.projectRootMarkers,
+        \ self.filesystemRootMarkers
+        \ )
+
+  let l:choices  = [
+          \ l:message,
+          \ printf('1. manual (default "%s")', l:initialDirectory),
+          \ printf('2. use initial working directory: "%s"', self.initialCwd)
+          \ ]
+
+  if v:null != l:rootDirByMarker
+    let l:item = '3. autodetected by root marker: '.l:rootDirByMarker
+    call add(l:choices, l:item)
+  endif
+
+  let l:choice = v:null
+  while index([0,1,2,3], l:choice) < 0
```

Comment:
It might be better to make it dynamic:
```suggestion
  while index(range(1, len(l:choices) - 1), l:choice) < 0
```

Reply:
At the moment it is better. I hope that it will be possible to get rid of these internals for any better input list implementation.

Reply:
[Tlib](https://github.com/vim-scripts/tlib/blob/master/autoload/tlib/input.vim) input list can be good example, fzf maybe also, but we have good piece of code at the phpactor side that to reuse after small modification.

Reply:
Also @dantleech prefer to avoid coupling with any other plugin or library.
And it's a good decision, so I would rather see us defining an interface contract.
I think there is already something close to that in `phpactor#input` for the lists.
Then in our side we only deal with what vim offer us `input`, `inputlist`, `confirm` and all that.
It's uggly and all but it's the same for anyone no matter what.

Then we could make some external plugins like `phpactor-vim-list-fzf`, that will only change the configuration variable so that phpactor vimscript will use it's own implementation for the list.

Reply:
We do of course have our `phpactor#input#` namespace...

Reply:
While I supposed to merge it in phpactor, I had not idea where to put such selector if it does not conform with (partially abstract) `phpactor#input#list` or any other. I deferred it in favor of resolving another issues.